pkg/controller/data_upload_controller.go

```diff
@@ -0,0 +1,498 @@
+	if result.Backup.EmptySnapshot {
+		du.Status.Message = "volume was empty so no data was upload"
+	}
+
+	if err := r.client.Patch(ctx, &du, client.MergeFrom(original)); err != nil {
+		log.WithError(err).Error("error updating DataUpload status")
+	}
+	log.Info("Data upload completed")
+}
+
+func (r *DataUploadReconciler) OnDataUploadFailed(ctx context.Context, namespace string, duName string, err error) {
+	defer r.closeDataPath(ctx, duName)
+
+	log := r.logger.WithField("dataupload", duName)
+
+	log.WithError(err).Error("Async fs backup data path failed")
+
+	var du velerov2alpha1api.DataUpload
+	if getErr := r.client.Get(ctx, types.NamespacedName{Name: duName, Namespace: namespace}, &du); getErr != nil {
+		log.WithError(getErr).Warn("Failed to get dataupload on failure")
+	} else {
+		if _, errOut := r.errorOut(ctx, &du, err, "data path backup failed", log); err != nil {
+			log.WithError(err).Warnf("Failed to patch dataupload with err %v", errOut)
+		}
+	}
+}
+
+func (r *DataUploadReconciler) OnDataUploadCancelled(ctx context.Context, namespace string, duName string) {
+	defer r.closeDataPath(ctx, duName)
+
+	log := r.logger.WithField("dataupload", duName)
+
+	log.Warn("Async fs backup data path canceled")
+
+	var du velerov2alpha1api.DataUpload
+	if getErr := r.client.Get(ctx, types.NamespacedName{Name: duName, Namespace: namespace}, &du); getErr != nil {
+		log.WithError(getErr).Warn("Failed to get dataupload on cancel")
+	} else {
+		original := du.DeepCopy()
+		du.Status.Phase = velerov2alpha1api.DataUploadPhaseCanceled
+		if du.Status.StartTimestamp.IsZero() {
+			du.Status.StartTimestamp = &metav1.Time{Time: r.clock.Now()}
+		}
+		du.Status.CompletionTimestamp = &metav1.Time{Time: r.clock.Now()}
+		if err := r.client.Patch(ctx, &du, client.MergeFrom(original)); err != nil {
+			log.WithError(err).Error("error updating DataUpload status")
+		}
+	}
+}
+
+func (r *DataUploadReconciler) OnDataUploadProgress(ctx context.Context, namespace string, duName string, progress *uploader.Progress) {
+	log := r.logger.WithField("dataupload", duName)
+
+	var du velerov2alpha1api.DataUpload
+	if err := r.client.Get(ctx, types.NamespacedName{Name: duName, Namespace: namespace}, &du); err != nil {
+		log.WithError(err).Warn("Failed to get dataupload on progress")
+		return
+	}
+
+	original := du.DeepCopy()
+	du.Status.Progress = shared.DataMoveOperationProgress{TotalBytes: progress.TotalBytes, BytesDone: progress.BytesDone}
+
+	if err := r.client.Patch(ctx, &du, client.MergeFrom(original)); err != nil {
+		log.WithError(err).Error("Failed to update progress")
+	}
+}
+
+// SetupWithManager registers the DataUpload controller.
+func (r *DataUploadReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&velerov2alpha1api.DataUpload{}).
+		Watches(&source.Kind{Type: &corev1.Pod{}}, kube.EnqueueRequestsFromMapUpdateFunc(r.findDataUploadForPod),
```

Comment:
Better to have some comments to explain why setting up the Manager like this.

Reply:
Added